src/esp/physics/bullet/BulletPhysicsManager.cpp

```diff
@@ -272,13 +310,93 @@ int BulletPhysicsManager::getNumActiveContactPoints() {
         static_cast<const btCollisionObject*>(manifold->getBody1());
 
     // logic copied from btSimulationIslandManager::buildIslands. We want to
-    // count manifold points only if related to non-sleeping bodies.
+    // count manifolds only if related to non-sleeping bodies.
     if (((colObj0) && colObj0->getActivationState() != ISLAND_SLEEPING) ||
         ((colObj1) && colObj1->getActivationState() != ISLAND_SLEEPING)) {
-      pointCount += manifold->getNumContacts();
+      count += manifold->getNumContacts();
+    }
+  }
+  return count;
+}
+
+std::vector<ContactPointData> BulletPhysicsManager::getContactPoints() const {
+  if (m_recentNumSubStepsTaken != 1) {
+    if (m_recentNumSubStepsTaken == -1) {
+      LOG(WARNING) << "getContactPoints: no previous call to stepPhysics";
+    } else {
+      // todo: proper logging-throttling API
+      static int count = 0;
+      if (count++ < 5) {
+        LOG(WARNING)
+            << "getContactPoints: the previous call to stepPhysics performed "
+            << m_recentNumSubStepsTaken
+            << " substeps, so getContactPoints's behavior may be unexpected.";
+      }
+      if (count == 5) {
+        LOG(WARNING)
+            << "getContactPoints: additional warnings will be suppressed.";
+      }
+    }
+  }
+
+  std::vector<ContactPointData> contactPoints;
+
+  // sloppy: assume fixedTimeStep_ hasn't changed since last call to stepPhysics
+  const float recentSubstepDt = fixedTimeStep_;
```

Comment:
Just to be clear on the bug described by this comment: if you call stepPhysics, then change fixedTimeStep_, then call getContactPoints, quantities like normalForce will be wrong. It would be accurate if we cached the actual recent substepDt somewhere.